android/src/main/java/com/swmansion/reanimated/nodes/ProceduralNode.java

```diff
@@ -0,0 +1,144 @@
+          arg.matchNodeWithOldContext(
+                  inputNode,
+                  previousEvalContext
+          );
+        }
+      } else if (mOldContext != previousEvalContext) {
+        throw new IllegalArgumentException("Tried to evaluate perform node in more than one contexts");
+      }
+
+      ProceduralNode proceduralNode = mNodesManager.findNodeById(mProceduralNode, ProceduralNode.class);
+      return proceduralNode.value(mEvalContext);
+    }
+  }
+
+  static public class ArgumentNode extends ValueNode {
+    private final SparseArray<Node> mValuesByContext = new SparseArray<>();
+    private final SparseArray<EvalContext> mContextsByValue = new SparseArray<>();
+    private final SparseArray<EvalContext> mOldContextByValue = new SparseArray<>();
+
+    public ArgumentNode(int nodeID, ReadableMap config, NodesManager nodesManager) {
+      super(nodeID, config, nodesManager);
+    }
+
+    public void matchContextWithNode(EvalContext context, Node node) {
+      mValuesByContext.put(context.contextID, node);
+      mContextsByValue.put(node.mNodeID, context);
+    }
+
+    public void dropContext(EvalContext evalContext) {
+      Node relatedNode = mValuesByContext.get(evalContext.contextID);
+      mContextsByValue.remove(relatedNode.mNodeID);
+      mValuesByContext.remove(evalContext.contextID);
+      mOldContextByValue.remove(relatedNode.mNodeID);
+    }
+
+    public void matchNodeWithOldContext(Node node, EvalContext evalContext) {
+      mOldContextByValue.put(node.mNodeID, evalContext);
+    }
+
+    @Override
+    public EvalContext contextForUpdatingChildren(EvalContext evalContext, Node lastVisited) {
+      if (lastVisited == null) {
+        return evalContext;
+      }
+      return mContextsByValue.get(lastVisited.mNodeID);
+    }
+
+    @Override
+    public void setValue(Object value, EvalContext context) {
+      ((ValueNode)mValuesByContext.get(context.contextID)).setValue(value, mNodesManager.globalEvalContext);
+    }
+
+    @Override
+    protected Object evaluate(EvalContext evalContext) {
+      if (evalContext == mNodesManager.globalEvalContext) {
+        throw new IllegalArgumentException("Tried to evaluate argumentNode in global context");
+      }
+      Node value = mValuesByContext.get(evalContext.contextID);
+      return value.doubleValue(mOldContextByValue.get(value.mNodeID));
+    }
+  }
+
+  private final int mResultNode;
+  private final int[] mProceduralArguments;
+
+  public ProceduralNode(int nodeID, ReadableMap config, NodesManager nodesManager) {
+    super(nodeID, config, nodesManager);
+    mResultNode = config.getInt("result");
+    mProceduralArguments = Utils.processIntArray(config.getArray("arguments"));
+  }
+
+  @Override
+  public @Nullable List<Node> filterChildrenByContext(EvalContext context) {
+    if (context.parent != null) {
+      List<Node> result = new ArrayList<>();
+      result.add(context.parent);
+      return result;
+    }
+    return mChildren;
```

Comment:
in what circumstances can we run this line over here? what the children are going to look like?

IMO it seem like this isn't the right method to inject the logic we want. If there is `context.parent` we just return an array of one object. Maybe it would be worth adding some additional logic to `findAndUpdateNodes` that'd handle the case where we should also visit `context.parent`. We already check for `ProceduralNode` there so it'd be much cleaner instead of adding new method that we only override in one place and create an array of only one item